src/game_api/thread_utils.hpp

```diff
@@ -1,57 +1,120 @@
 #pragma once
 
-#include <Windows.h> // for ResumeThread, SuspendThread, HANDLE
-#include <cstddef>   // for size_t
-#include <cstdint>   // for int64_t
+#include <cstddef>  // for size_t
+#include <cstdint>  // for int64_t
+#include <stdlib.h> // for free
 
-HANDLE get_main_thread();
-size_t heap_base();
-size_t local_heap_base();
+struct PRNG;
+struct StateMemory;
+struct LevelGenSystem;
+struct LiquidPhysics;
 
-// Used for objects that are allocated with the game's custom allocator
-template <typename T>
-class OnHeapPointer
+struct HeapBase
 {
-    int64_t ptr_;
+    // get HeapBase from save slots
+    static HeapBase get(uint8_t slot);
+    // get local, fallback to main if can't get local
+    static HeapBase get();
+    // use only if you know what you're doing
+    static HeapBase get_main();
 
-  public:
-    explicit OnHeapPointer(size_t ptr)
-        : ptr_(ptr)
+    bool is_null() const noexcept
     {
+        return ptr == NULL;
     }
+    uintptr_t address() const noexcept
+    {
+        return ptr;
+    }
+    StateMemory* state() const noexcept
+    {
+        if (is_null())
+            return nullptr;
 
-    T* decode() const
+        return reinterpret_cast<StateMemory*>(ptr + GAME_OFFSET::STATE);
+    }
+    uint32_t frame_count() const noexcept
     {
-        return reinterpret_cast<T*>(ptr_ + heap_base());
+        if (is_null())
+            return NULL;
+
+        return *reinterpret_cast<uint32_t*>(ptr + GAME_OFFSET::FRAME_COUNTER);
     }
+    PRNG* prng() const noexcept
+    {
+        if (is_null())
+            return nullptr;
 
-    T* decode_local() const
+        return reinterpret_cast<PRNG*>(ptr + GAME_OFFSET::_PRNG);
+    }
+    LevelGenSystem* level_gen() const noexcept
     {
-        auto lhb = local_heap_base();
-        if (lhb == 0)
+        if (is_null())
             return nullptr;
 
-        return reinterpret_cast<T*>(ptr_ + lhb);
+        return reinterpret_cast<LevelGenSystem*>(ptr + GAME_OFFSET::LEVEL_GEN);
     }
+    LiquidPhysics* liquid_physics() const noexcept
+    {
+        if (is_null())
+            return nullptr;
 
-    T* operator->() const
+        return reinterpret_cast<LiquidPhysics*>(ptr + GAME_OFFSET::LIQUID_ENGINE);
+    }
+
+    void copy_to(HeapBase other) const;

```

Comment:
Was also thinking about making static function with two arguments. Should i or is this clear enough?

Reply:
any of them seem clear enough to me